File: verdant-path/src/verdant_path/tracker.py
"""Daily check-ins, workout logs, and journal entries.

These are the member-facing records: what they did, how they felt, and what they
noticed. They feed fatigue scoring, TRIMP totals, and weekly reviews.
"""

from __future__ import annotations

from dataclasses import dataclass, field
from datetime import date

from verdant_path.fatigue import fatigue_cue, fatigue_from_checkin
from verdant_path.movements import Movement
from verdant_path.trimp import TRIMP, trimp_for_session


@dataclass(frozen=True)
class SetLog:
    """A single logged set: the raw unit of a training session."""

    load: float | None = None
    reps: int | None = None
    tempo: str | None = None       # e.g. "3010"
    rpe: float | None = None       # 1-10 perceived exertion
    rest_sec: int | None = None
    notes: str | None = None


@dataclass
class WorkoutLog:
    """A member's completed session: movements, sets, and resulting TRIMP."""

    day: date
    movement: Movement
    sets: list[SetLog] = field(default_factory=list)
    trimp: TRIMP | None = None          # explicit override after the session
    notes: str | None = None

    def session_trimp(self, assigned_level: TRIMP | None = None) -> TRIMP:
        """Resolve the session's TRIMP: explicit override, else estimate."""

        if self.trimp is not None:
            return self.trimp
        return trimp_for_session(self.sets, assigned_level=assigned_level)


@dataclass
class CheckIn:
    """A daily check-in: the five signals that produce a fatigue score."""

    day: date
    energy: float          # 1-5
    mood: float            # 1-5
    soreness: float        # 1-5 (5 = very sore)
    sleep_hours: float     # hours
    hrv: float             # ms
    hrv_baseline: float = 50.0

    def fatigue_percent(self) -> float:
        return fatigue_from_checkin(
            self.energy,
            self.mood,
            self.soreness,
            self.sleep_hours,
            self.hrv,
            hrv_baseline=self.hrv_baseline,
        )

    def cue(self):
        return fatigue_cue(self.fatigue_percent())


@dataclass
class JournalEntry:
    """A free-form reflective note: pain, energy, mood, insights.

    Journaling is introspection — the embodied-awareness layer that gives the
    metrics their meaning. Tags keep entries retrievable over time.
    """

    day: date
    text: str
    tags: list[str] = field(default_factory=list)
    photo: str | None = None          # path or URI to a progress photo

    def __post_init__(self) -> None:
        # Normalize tags on construction so callers can pass "#Recovery" etc.
        normalized: list[str] = []
        for tag in self.tags:
            clean = tag.strip().lstrip("#").lower()
            if clean and clean not in normalized:
                normalized.append(clean)
        self.tags = normalized

    def add_tag(self, tag: str) -> None:
        clean = tag.strip().lstrip("#").lower()
        if clean and clean not in self.tags:
            self.tags.append(clean)

File: verdant-path/mobile/src/domain/tracker.ts
/**
 * Daily check-ins, workout logs, and journal entries.
 *
 * These are the member-facing records: what they did, how they felt, and what
 * they noticed. They feed fatigue scoring, TRIMP totals, and weekly reviews.
 */

import { Fatigue, fatigueCue, fatigueFromCheckin } from "./fatigue";
import { Movement } from "./movements";
import { Trimp, trimpForSession, type SetLog } from "./trimp";

export type { SetLog };

export interface CheckIn {
  id: string;
  day: Date;
  energy: number; // 1-5
  mood: number; // 1-5
  soreness: number; // 1-5 (5 = very sore)
  sleepHours: number; // hours
  hrv: number; // ms
  hrvBaseline?: number; // ms
  readonly fatiguePercent: number;
  readonly cue: Fatigue;
}

export function makeCheckIn(args: Omit<CheckIn, "fatiguePercent" | "cue">): CheckIn {
  const fatiguePercent = fatigueFromCheckin({
    energy: args.energy,
    mood: args.mood,
    soreness: args.soreness,
    sleepHours: args.sleepHours,
    hrv: args.hrv,
    hrvBaseline: args.hrvBaseline,
  });
  return { ...args, fatiguePercent, cue: fatigueCue(fatiguePercent) };
}

export interface WorkoutLog {
  id: string;
  day: Date;
  movement: Movement;
  sets: SetLog[];
  trimp?: Trimp; // explicit override after the session
  notes?: string;
}

/** Resolve the session's TRIMP: explicit override, else estimate from sets. */
export function sessionTrimpFor(log: WorkoutLog): Trimp {
  if (log.trimp !== undefined) return log.trimp;
  return trimpForSession(log.sets);
}

export interface JournalEntry {
  id: string;
  day: Date;
  text: string;
  tags: string[];
  photo?: string; // path or URI to a progress photo
}

/** Normalize a raw tag string ("#Recovery", "  #soreness ") to "recovery". */
export function normalizeTag(tag: string): string {
  return tag.trim().replace(/^#+/, "").toLowerCase();
}

/** Build a JournalEntry with normalized, de-duplicated tags. */
export function makeJournalEntry(args: {
  id: string;
  day: Date;
  text: string;
  tags?: string[];
  photo?: string;
}): JournalEntry {
  const seen = new Set<string>();
  const tags: string[] = [];
  for (const raw of args.tags ?? []) {
    const clean = normalizeTag(raw);
    if (clean && !seen.has(clean)) {
      seen.add(clean);
      tags.push(clean);
    }
  }
  return { id: args.id, day: args.day, text: args.text, tags, photo: args.photo };
}
